pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,278 @@
+			addedSome = true
+		}
+	}
+
+	// add new statuscollector data
+	for statusCollectorName, statusCollectorSpec := range statusCollectorNameToSpec {
+		if _, ok := c.statusCollectorNameToData[statusCollectorName]; !ok {
+			c.statusCollectorNameToData[statusCollectorName] = &statusCollectorData{
+				StatusCollectorSpec: statusCollectorSpec,
+				workStatusRefToData: make(map[string]*workStatusData),
+			}
+
+			addedSome = true
+		}
+	}
+
+	return removedSome, addedSome
+}
+
+// updateStatusCollector updates the status collector data in the
+// combinedstatus resolution. If the status collector is not relevant to the
+// latter, the function returns false. The function returns true if the status
+// collector data is updated. The given spec is assumed to be valid and
+// immutable.
+func (c *combinedStatusResolution) updateStatusCollector(statusCollectorName string,
+	statusCollectorSpec *v1alpha1.StatusCollectorSpec) bool {
+	c.Lock()
+	defer c.Unlock()
+
+	scData, ok := c.statusCollectorNameToData[statusCollectorName]
+	if !ok {
+		return false // statusCollector is irrelevant to this combinedstatus resolution
+	}
+
+	if statusCollectorSpecsMatch(scData.StatusCollectorSpec, statusCollectorSpec) {
+		return false // statusCollector data is already up-to-date
+	}
+
+	// status collector clauses need to be updated, therefore update fields
+	// and invalidate all cached workstatus evaluations by resetting the map
+	c.statusCollectorNameToData[statusCollectorName] = &statusCollectorData{
+		StatusCollectorSpec: statusCollectorSpec,
+		workStatusRefToData: make(map[string]*workStatusData),
+	}
+
+	return true
+}
+
+// evaluateWorkStatus evaluates the workstatus per relevant statuscollector.
+// The function returns true if any statuscollector data is updated.
+func (c *combinedStatusResolution) evaluateWorkStatus(ctx context.Context, celEvaluator *celEvaluator,
+	workStatusWECName string, workStatusContent *runtime.RawExtension) (bool, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	logger := klog.FromContext(ctx)
+
+	updated := false
+	for statusCollectorName, scData := range c.statusCollectorNameToData {
+		changed, err := evaluateWorkStatusAgainstStatusCollector(celEvaluator, workStatusWECName, workStatusContent,
+			scData)
+		if err != nil {
+			logger.Error(err, "failed to evaluate workstatus against statuscollector",
+				"statusCollectorName", statusCollectorName)
+			return false, nil
+		}
+
+		updated = updated || changed
+	}
+
+	return updated, nil
+}
+
+// evaluateWorkStatusAgainstStatusCollector evaluates the workstatus against
+// the statuscollector clauses and caches the evaluations. If the workstatus
+// does not match the filter, no other clause is evaluated.
+// The function returns true if an evaluation is updated.
+// If any evaluation fails, the function returns an error.
+// The function assumes that the caller holds a lock over the combinedstatus
```

Comment:
I usually use a naming convention to make locking discipline easier to check. The name of a method that is called with the lock held ends with "Locked".